fix(compiler): ensure that placeholders have the correct sourceSpan

petebacondarwin opened the pull request:
When the `preserveWhitespaces` is not true, the template parser will
process the parsed AST nodes to remove excess whitespace. Since the
generated `goog.getMsg()` statements rely upon the AST nodes after
this whitespace is removed, the i18n extraction must make a second pass.

Previously this resulted in innacurrate source-spans for the i18n text and
placeholder nodes that were extracted in the second pass.

This commit fixes this by reusing the source-spans from the first pass
when extracting the nodes in the second pass.

Fixes #39671

petebacondarwin committed:
fix(compiler): ensure that placeholders have the correct sourceSpan

When the `preserveWhitespaces` is not true, the template parser will
process the parsed AST nodes to remove excess whitespace. Since the
generated `goog.getMsg()` statements rely upon the AST nodes after
this whitespace is removed, the i18n extraction must make a second pass.

Previously this resulted in innacurrate source-spans for the i18n text and
placeholder nodes that were extracted in the second pass.

This commit fixes this by reusing the source-spans from the first pass
when extracting the nodes in the second pass.

Fixes #39671

AndrewKushnir commented on packages/compiler/src/i18n/i18n_parser.ts:
```diff
@@ -105,12 +105,13 @@ class _I18nVisitor implements html.Visitor {
 
   visitAttribute(attribute: html.Attribute, context: I18nMessageVisitorContext): i18n.Node {
-    const node = this._visitTextWithInterpolation(attribute.value, attribute.sourceSpan, context);
+    const node = this._visitTextWithInterpolation(
+        attribute.value, attribute.valueSpan || attribute.sourceSpan, context, attribute.i18n);
     return context.visitNodeFn(attribute, node);
```
The logic inside the `_visitTextWithInterpolation` function is a bit tricky. May be we can put it here instead?

```suggestion
    const node = this._visitTextWithInterpolation(
        attribute.value, attribute.valueSpan || attribute.sourceSpan, context);
    if (attribute.i18n) {
      // ... comments here ...
      reuseExistingSourceSpans(node, attribute.i18n.nodes[0]);
    }
    return context.visitNodeFn(attribute, node);
```
and with `visitText`:

```suggestion
  visitText(text: html.Text, context: I18nMessageVisitorContext): i18n.Node {
    const node = this._visitTextWithInterpolation(text.value, text.sourceSpan, context);
    if (text.i18n) {
      // ... comments here ...
      reuseExistingSourceSpans(node, text.i18n);
    }    
    return context.visitNodeFn(text, node);
  }
```
What do you think?

petebacondarwin replied:
Thanks for the review @AndrewKushnir. I agree that `_visitTextWithInterpolation()` is becoming too complex.
Rather than move the call outside of the method, though, I have refactored that method to be more maintainable.

petebacondarwin replied:
I did the refactoring in a separate commit, so that it is easier to see the actual fix for the issue.

AndrewKushnir replied:
Thanks Pete! 👍

petebacondarwin committed:
refactor(compiler): tidy up interpolation splitting

When parsing for i18n messages, interpolated strings are
split into `Text` and `Placeholder` pieces.  The method that
does this `_visitTextWithInterpolation()` was becoming too
complex. This commit refactors that method along with some
associated functions that it uses.

AndrewKushnir reviewed (approved):
Thanks for the cleanup and the fix @petebacondarwin 👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:342901543:BASE:342901566:1605639959811:2db1ff18).

AndrewKushnir commented:
@petebacondarwin FYI, presubmit is successful for the changes in this PR. Thank you.

petebacondarwin committed:
fixup! fix(compiler): ensure that placeholders have the correct sourceSpan

petebacondarwin committed:
test(compiler-cli): add source-mapping test helper

This helper improves the message given when an expectation
fails in a source-mapping test.

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:342901543:BASE:343373044:1605825654270:189ae8f).
